File: swig/ruby/test_hash_counter.rb
require 'minitest/autorun'
require 'jellyfish'

class TestHashCounter < MiniTest::Unit::TestCase
  def setup
    Jellyfish::MerDNA::k(100)
    @hash = Jellyfish::HashCounter.new(1024, 5)
  end

  def test_info
    assert_equal(100, Jellyfish::MerDNA::k)
    assert_equal(1024, @hash.size)
    assert_equal(5, @hash.val_len)
#    assert_equal(1, @hash.nb_threads)
  end

  def test_add
    mer = Jellyfish::MerDNA.new

    1000.times { |i|
      mer.randomize!
      val = rand(1000)
      assert(@hash.add(mer, val))
      if i % 3 > 0
        nval = rand(1000)
        val += nval
        if i % 3 == 1
          assert(!@hash.add(mer, nval))
        else
          assert(@hash.update_add(mer, nval))
        end
      end
      assert_equal(val, @hash.get(mer))
      assert_equal(val, @hash[mer])
    }

    mer.randomize!
    assert_nil(@hash.get(mer))
    assert_nil(@hash[mer])
    assert(!@hash.update_add(mer, 1))
  end
end

File: swig/python/test_hash_counter.py
import unittest
import sys
import random
import jellyfish

class TestHashCounter(unittest.TestCase):
    def setUp(self):
        jellyfish.MerDNA.k(100)
        self.hash = jellyfish.HashCounter(1024, 5)

    def test_info(self):
        self.assertEqual(100, jellyfish.MerDNA.k())
        self.assertEqual(1024, self.hash.size())
        self.assertEqual(5, self.hash.val_len())

    def test_add(self):
        mer  = jellyfish.MerDNA()
        good = True
        for i in range(1000):
            mer.randomize()
            val = random.randrange(1000)
            good = good and self.hash.add(mer, val)
            if not good: break
            if i % 3 > 0:
                nval = random.randrange(1000)
                val  = val + nval
                if i % 3 == 1:
                    good = good and (not self.hash.add(mer, nval))
                else:
                    good = good and self.hash.update_add(mer, nval)
            if not good: break
            good = good and (val == self.hash.get(mer)) and (val == self.hash[mer])
            if not good: break
        self.assertTrue(good)


if __name__ == '__main__':
    data = sys.argv.pop(1)
    unittest.main()
